### DISCUSSION TITLE
v3.5.0-M4

### DISCUSSION THREAD

**Discussion by nafg** [slick/slick#2759] [Announcements]
### Highlights 🎉
* Revert drop scala 212 by @mdedetrich in https://github.com/slick/slick/pull/2734
* #2701 Add conditional filterNot methods (filterNotIf/filterNotOpt) support by @sebastianrozyczka in https://github.com/slick/slick/pull/2751
* Scala 3 support by @nafg in https://github.com/slick/slick/pull/2687
### Chores 🛠
* Update dependency org.xerial.sbt:sbt-sonatype to v3.9.19 by @renovate in https://github.com/slick/slick/pull/2712
* Update sbt-sonatype to 3.9.19 by @slick-scala-steward in https://github.com/slick/slick/pull/2713
* Update sbt-ci-release to 1.5.12 by @slick-scala-steward in https://github.com/slick/slick/pull/2715
* Update dependency org.xerial.sbt:sbt-sonatype to v3.9.20 by @renovate in https://github.com/slick/slick/pull/2716
* Update sbt to 1.8.3 by @slick-scala-steward in https://github.com/slick/slick/pull/2718
* Update dependency org.scalatestplus:testng-7-5 to v3.2.16.0 by @renovate in https://github.com/slick/slick/pull/2721
* Update sbt-sonatype to 3.9.21 by @slick-scala-steward in https://github.com/slick/slick/pull/2722
* Update ojdbc8_g to 21.10.0.0 by @slick-scala-steward in https://github.com/slick/slick/pull/2723
* Update sbt to 1.9.0 by @slick-scala-steward in https://github.com/slick/slick/pull/2728
* Update dependency org.scoverage:sbt-scoverage to v2.0.8 by @renovate in https://github.com/slick/slick/pull/2730
* Update sbt-scoverage to 2.0.8 by @slick-scala-steward in https://github.com/slick/slick/pull/2729
* Update dependency ch.epfl.scala:sbt-version-policy to v2.1.1 by @renovate in https://github.com/slick/slick/pull/2732
* Update scala-compiler, scala-library to 2.13.11 by @slick-scala-steward in https://github.com/slick/slick/pull/2737
* Update logback-classic to 1.3.8 by @slick-scala-steward in https://github.com/slick/slick/pull/2740
* Update scala-compiler, scala-library to 2.12.18 by @slick-scala-steward in https://github.com/slick/slick/pull/2741
* Update actions/upload-artifact action to v3 by @renovate in https://github.com/slick/slick/pull/2744
* Update dependency org.scala-lang.modules:scala-collection-compat to v2.11.0 by @renovate in https://github.com/slick/slick/pull/2749
* Update scala-collection-compat to 2.11.0 by @slick-scala-steward in https://github.com/slick/slick/pull/2748
* Update dependency sbt/sbt to v1.9.1 by @renovate in https://github.com/slick/slick/pull/2752
* Update mysql client dependency by @xuwei-k in https://github.com/slick/slick/pull/2754
* Update sbt to 1.9.2 by @slick-scala-steward in https://github.com/slick/slick/pull/2757
### Other Changes
* A few cleanups by @nafg in https://github.com/slick/slick/pull/2717
* Remove headOption extension method for ReadAheadIterator by @nafg in https://github.com/slick/slick/pull/2719
* Remove usage of reflection to support pre-JDBC 4.1 (pre-Java 7) by @nafg in https://github.com/slick/slick/pull/2720
* Some cleanup by @nafg in https://github.com/slick/slick/pull/2739

## New Contributors
* @mdedetrich made their first contribution in https://github.com/slick/slick/pull/2734
* @sebastianrozyczka made their first contribution in https://github.com/slick/slick/pull/2751

**Full Changelog**: https://github.com/slick/slick/compare/v3.5.0-M3...v3.5.0-M4

<hr /><em>This discussion was created from the release <a href='https://github.com/slick/slick/releases/tag/v3.5.0-M4'>v3.5.0-M4</a>.</em>

**Comment by xuwei-k** [comment]
Thank you for release with Scala 3 support🎉 👍 

reported an issue 🙏

- https://github.com/slick/slick/issues/2761

**Comment by He-Pin** [comment]
Thank you ~~

**Comment by mkurz** [comment]
Any ETA for 3.5.0 final release?

**Comment by nafg** [comment]
I've been very busy but as always, if someone volunteers to "pair program" or just do a session watching me code, I will allocate time.

The problem right now is making sure that bringing downstream libraries to Scala 3 won't require a breaking change in Slick, and we got stuck migrating them.

But on second thought, a library can be updated to Slick 3.5.0 without introducing Scala 3 cross-compiling, so we could skip it.

But it seems getting slick-additions to compile with Scala 3 might indeed require a breaking change in Slick (I need to investigate better). It would be very unfortunate if users of slick-additions or slick-pg can't upgrade to Scala 3 until Slick 3.6.0 (which could be as soon as needed, but still annoying). So it's not pointless.

@mdedetrich was working on migrating slick-pg but I don't know the status. I could use help with slick-additions either way.

**Reply by anqit** [reply]
Hey @nafg, I believe you and I did this at some point last year, but I would love to jump on a call at some point and try to get a better idea about Slick's innards and if lucky, maybe even help out ;)

**Reply by nafg** [reply]
Great! Ping me on Discord and we can coordinate.

**Reply by mdedetrich** [reply]
I am a quite busy (in the middle of renovating) so unfortunately my capacity is going to be quite limited this month but if something critical comes up I can try make some time,

From my (i.e. Pekko's) end the most critical thing which really should be a blocker is https://github.com/slick/slick/issues/2811 (the reason for the "blocker" status is quite obvious, Slick appears to have changed behaviour somehow although this also needs to be validated).

There is also the slick-pg work that @nafg mentioned, i.e. https://github.com/tminglei/slick-pg/pull/668. The end goal is to get scala3 support for slick-pg but the complication here is that slick-pg uses reflection. While in a lot of cases the reflection is trivial (i.e. getting class name from a class) there is one case that is slightly more complicated and there are open questions on which reflection solution to use (there might even be a question of whether slick core should adopt something like izumi-reflect so that downstream libraries like slick-additions/slick-pg have a common base for reflection like capabilities).

**Reply by nafg** [reply]
> But it seems getting slick-additions to compile with Scala 3 might indeed require a breaking change in Slick

MappedProjection's second parameter was removed in M5 for this reason (and because AFAICT it's unnecessary)

**Comment by nafg** [comment]
It seems that now Slick-additions now fully compiles for Scala 3 with Slick 3.5.0-M5, but there's one test that fails, and it could be a bug in Slick.

See https://github.com/nafg/slick-additions/pull/272 

It actually seems to be similar to @mdedetrich 's issue IIRC, in that it's an insert that isn't showing up.

**Reply by nafg** [reply]
Uhh never mind the issue is ScalaTest's definition of `===` is overriding Slick's

**Comment by nafg** [comment]
Also, @hughsimpson seems to have ported slick-pg to Scala 3 in https://github.com/tminglei/slick-pg/pull/677. The issue there is that I don't know if @tminglei is around to merge and release it.

If there is no response at some point I should probably fork slick-pg under the Slick organization and release it myself, as per https://github.com/tminglei/slick-pg/issues/514#issuecomment-1001262647
